ext/mysqlnd/mysqlnd_structs.h

```diff
@@ -290,7 +290,7 @@ typedef struct st_mysqlnd_unbuffered_result MYSQLND_RES_UNBUFFERED;
 typedef struct st_mysqlnd_debug MYSQLND_DEBUG;
 
 
-typedef MYSQLND_RES* (*mysqlnd_stmt_use_or_store_func)(MYSQLND_STMT * const);
+typedef enum_func_status (*mysqlnd_stmt_use_or_store_func)(MYSQLND_STMT * const);
```

Comment:
Why these changes? Isn't failure determined by non-null return?

We will not be able to merge this to stable releases with this change.

Reply:
In mysqlnd.h there was a ternary operator that I removed which was casting this value to a boolean. 

```
#define mysqlnd_stmt_store_result(stmt)	(!mysqlnd_stmt_field_count((stmt)) ? PASS:((stmt)->m->store_result((stmt))? PASS:FAIL))
```

As far as I am aware the actual result of these functions was never exposed to mysqli or PDO_mysql. The result was meant to be stored inside of the object property so returning it didn't make much sense. If you grep for `mysql_stmt_store_result` you will see that the only time this function is used is in boolean context. Same with `default_rset_handler` which is only ever used in void context. 
I agree that we could keep it as a nullable object type and then immediately cast it to boolean but why? What purpose does that serve? I am genuinely asking because I am no C experts so I don't know. Is there any memory consideration? Are internals of mysqlnd exposed to any other extensions? If there's no reason then I think it is cleaner to let that function return PASS/FAIL instead of using that ternary operator.  

What does it mean that we will not be able to merge it into stable releases? Is there something that makes this unstable?

Reply:
> Are internals of mysqlnd exposed to any other extensions?

Yes, unfortunately. For this reason mysqlnd APIs are subject to ABI stability guarantees. We can make this change on master, but we cannot make it on PHP-7.4.

Reply:
It's fine with me to only push it to PHP 8. We can make a separate patch for PHP 7.4 where we only set the `stmt->error_info` and add `DBG_RETURN(NULL);`.
If you decide that it is better to leave the method signature alone, then we can do that too. As I said mysqli and PDO don't care about that function as they never access it directly. 
Let me know what you decide. I can roll back that change and add the ternary back in.

Reply:
I would prefer to first fix the issue and then do refactoring on master. It will be hard to merge this across branches otherwise. The diff currently also contains lots of unrelated changes that make it hard to see what you're actually changing. E.g. all the changes in ext/mysqli seem unnecessary? Please avoid changing whitespace on unrelated lines.